### PR TITLE
Enhancement : Validations on survey edit page

### PR DESCRIPTION
**Pull request by Dhruwang**
## What does this PR do?

Initially there were no validations on survey edit page,
User was able to save questions with no headlines
Blank Multiple choice options were allowed 

https://github.com/formbricks/formbricks/assets/67850763/33f98085-6d84-4ae4-ad3d-9b838fc5732c

Screenshot from 2023-07-20 09-53-10

## Approach

Validations rules have been defined in an object names validationRules in **SurveyEditor.tsx**
More validations can be added by extending this object further
By default it will always check for question headline
```
const validationRules = {
    multipleChoiceMulti: (question: any) => {
      return !question.choices.some((element: any) => element.label.trim() === "");
    },
    multipleChoiceSingle: (question: any) => {
      return !question.choices.some((element: any) => element.label.trim() === "");
    },
    defaultValidation: (question: any) => {
      return question.headline.trim() !== "";
    },
  };
```
Whenever user tries to save the survey , it will undergo validation using 
```
const validateSurvey = (survey) => {
    faultyQuestions = [];
    for (let index = 0; index < survey.questions.length; index++) {
      const question = survey.questions[index];
      const isValid = validateQuestion(question);

      if (!isValid) {
        faultyQuestions.push(question.id);
      }
    }
    // if there are any faulty questions, the user won't be allowed to save the survey
    if (faultyQuestions.length > 0) {
      setInvalidQuestions(faultyQuestions);
      toast.error("Please fill required fields");
      return false;
    }

    return true;
  };

```
And if any of the question is invalid, its id will be pushed to a temporary array `faultyQuestions` which is further used to set value of `invalidQuestions` array 

`InvalidQuestions` array is then further used to determine the value of boolean `isFaulty` which is passed as a prop in `QuestionCard` component and for some question if isFaulty is true then its color changes to red

```
         <QuestionCard
                    key={internalQuestionIdMap[question.id]}
                    localSurvey={localSurvey}
                    questionIdx={questionIdx}
                    moveQuestion={moveQuestion}
                    updateQuestion={updateQuestion}
                    duplicateQuestion={duplicateQuestion}
                    deleteQuestion={deleteQuestion}
                    activeQuestionId={activeQuestionId}
                    setActiveQuestionId={setActiveQuestionId}
                    lastQuestion={questionIdx === localSurvey.questions.length - 1}
                    isFaulty={invalidQuestions ? invalidQuestions.includes(question.id) : false}
                  />
```

Once user has tried to save the invalid survey `invalidQuestions` with have some value and then onward it will validate the individual question as soon as user modifies them in order to remove them or keep them in `invalidQuestions` array

```
 const validateSurvey = (question: any) => {
    // prevent this function to execute further if user hasnt still tried to save the survey
    if (invalidQuestions === null) {
      return;
    }
    let temp = invalidQuestions;
    if (validateQuestion(question)) {
      temp = invalidQuestions.filter((id) => id !== question.id);
      setInvalidQuestions(temp);
      return;
    } else if (!invalidQuestions.includes(question.id)) {
      temp.push(question.id);
      setInvalidQuestions(temp);
    }
  };

```
Logic of validateQuestion 
```
const validateQuestion = (question) => {
  // Fetch specific validation function based on question type
  const specificValidation = validationRules[question.type];

  // Use default validation function as fallback
  const defaultValidation = validationRules.defaultValidation;

  // Check if specific validation exists for the question type, and run it
  const specificValidationResult = specificValidation ? specificValidation(question) : true;

  // Always run the default validation
  const defaultValidationResult = defaultValidation(question);

  // Return true only if both specific and default validation pass
  return specificValidationResult && defaultValidationResult;
};
```
## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [x] Enhancement (small improvements)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change adds a new database migration
- [ ] This change requires a documentation update

## How should this be tested?

1. Create a new survey or try to edit an existing survey
2. Leave some fields empty like Question's headline
3. Try to save

## Checklist

- [x] Added a screen recording or screenshots to this PR
- [x] Filled out the "How to test" section in this PR
- [x] Read the [contributing guide](https://github.com/formbricks/formbricks/blob/main/CONTRIBUTING.md)
- [x] Self-reviewed my own code
- [x] Commented on my code in hard-to-understand bits
- [x] Ran `pnpm build`
- [x] Checked for warnings, there are none
- [x] Removed all `console.logs`
- [x] Merged the latest changes from main onto my branch with `git pull origin main`
- [x] My changes don't cause any responsiveness issues
- [x] Updated the Formbricks Docs if changes were necessary

### CODE REVIEW

**Comment by Dhruwang** [comment]
Hey @jobenjada 

I have introduced a slight modification to the method. Previously, the `invalidQuestions` array stored `questionIdx`, which could lead to complications when users attempted to drag, move, or delete questions, as these actions could alter the order. Therefore, I have now updated it to store the `id` of the questions that are invalid instead.

File: apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/edit/RatingQuestionForm.tsx [hunk 1]
```diff
@@ -33,6 +35,7 @@ export default function RatingQuestionForm({
             name="headline"
             value={question.headline}
             onChange={(e) => updateQuestion(questionIdx, { headline: e.target.value })}
+            isInvalid={isFaulty && question.headline.trim() === ""}
```
Comment (mattinannt) [inline apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/edit/RatingQuestionForm.tsx:38]: why is the `&& question.headline.trim() === ""` needed? isn't this something the validation function should handle?

**Review by mattinannt — CHANGES_REQUESTED** [review]
@Dhruwang thanks a lot for the PR 😊 The red highlight when the question is invalid looks pretty cool 😊🎉🚀
I left a few comments, would be great if you can address these issues :-) After that I can test again and if everything still works great, we can merge the changes 😊

File: apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/edit/RatingQuestionForm.tsx (hunk 1 above)
Comment (Dhruwang) [reply apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/edit/RatingQuestionForm.tsx:38]: Responsibility for marking a question as faulty or invalid lies with `isFaulty` To identify the specific faulty field within that particular question, the following expression is employed: `isFaulty && question.headline.trim() === ""`.

It is important to note that if `isFaulty` is false, there is no need to proceed with further checks, as the question is deemed valid. However, in cases where the question is indeed faulty, this expression helps pinpoint which field is at fault.

**Comment by Dhruwang** [comment]
On it 😊

File: apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/edit/RatingQuestionForm.tsx (hunk 1 above)
Comment (mattinannt) [reply apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/edit/RatingQuestionForm.tsx]: Got it! :-)
The issue I am seeing is that we now have the validation rules in many different places. We define in `Validation.ts` when a question is invalid but additionally define rules in every input field when this is invalid. Ideally we can combine these to globally manage the validation rules.

Comment (mattinannt) [reply apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/edit/RatingQuestionForm.tsx]: Long-term I would prefer a more standardised field validation that works with e.g. react hook form ( https://react-hook-form.com/get-started#Applyvalidation ) which we also use in other places in the app.

The tricky part is that we have a form for every question and have the save/publish actions far on the outside of this code. But I think if we find a good solution there using react hook form would be a good approach to define validation rules and also work with error messages, etc.

**Review by mattinannt — APPROVED** [review]
@Dhruwang Looks great! 💪 I only added the types for the validation and now we are ready to merge the changes! 😊 Great job 🎉

**Pull request merged**